src/mlpack/methods/ann/sparse_autoencoder_function.hpp

```diff
@@ -0,0 +1,407 @@
+#ifndef __MLPACK_METHODS_ANN_SPARSE_AUTOENCODER_FUNCTION_HPP
+#define __MLPACK_METHODS_ANN_SPARSE_AUTOENCODER_FUNCTION_HPP
+
+#include <mlpack/core.hpp>
+
+#include <type_traits>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * This is a class for the sparse autoencoder objective function. It can be used
+ * to create learning models like self-taught learning, stacked autoencoders,
+ * conditional random fields (CRFs), and so forth.
+ * @code
+ * #include <mlpack/methods/ann/sparse_autoencoder_function.hpp>
+ * #include <mlpack/methods/ann/activation_functions/lazy_logistic_function.hpp>
+ *
+ * #include <mlpack/core.hpp>
+ * #include <mlpack/methods/sparse_autoencoder/sparse_autoencoder.hpp>
+ * #include <mlpack/methods/ann/layer/base_layer.hpp>
+ *
+ * #include <iostream>
+ *
+ * int main()
+ *  {
+ *   using namespace mlpack;
+ *   using namespace arma;
+ *
+ *   using FSigmoidLayer = ann::SigmoidLayer<ann::LazyLogisticFunction>;
+ *
+ *   using SAEF = ann::SparseAutoencoderFunction<FSigmoidLayer, FSigmoidLayer>;
+ *   //If you want to compare the performance with original class, replace
+ *   //SAEF with following definition
+ *   //using SAEF = nn::SparseAutoencoderFunction;
+ *
+ *   size_t const Features = 16*16;
+ *   arma::mat data = randu<mat>(Features, 10000);
+ *
+ *   SAEF encoderFunction(data, Features, Features / 2);
+ *   const size_t numIterations = 100; // Maximum number of iterations.
+ *   const size_t numBasis = 10;
+ *   optimization::L_BFGS<SAEF> optimizer(encoderFunction, numBasis, numIterations);
+ *
+ *   arma::mat parameters = encoderFunction.GetInitialPoint();
+ *
+ *   // Train the model.
+ *   Timer::Start("sparse_autoencoder_optimization");
+ *   const double out = optimizer.Optimize(parameters);
+ *   Timer::Stop("sparse_autoencoder_optimization");
```

Comment:
The API here is the way Siddharth originally wrote it, using the standard mlpack optimizers to optimize the weights of the network.  But maybe it would be a better idea to make this class work with the `Trainer` class in `src/mlpack/methods/ann/`?  Kind of like the examples in `convolutional_network_test.cpp` and `feedforward_network_test.cpp`:

```
SAEF autoencoder(...);
Trainer<SAEF> trainer(autoencoder, ...);
trainer.Train(...);
```

This would help make all of the ANN-related code in mlpack have a unified interface.  I'd be interested in zoq's comments on this too, since I don't know his plans for what the API there will eventually look like.